src/main.rs

```diff
@@ -140,120 +136,57 @@ async fn request(
-    PROVIDERS.with(|p| {
-        p.borrow()
-            .iter()
-            .map(|(_, provider)| provider.into())
-            .collect::<Vec<ProviderView>>()
-    })
-}
-
-#[update(name = "registerProvider", guard = "require_register_provider")]
-#[candid_method(rename = "registerProvider")]
-fn register_provider(provider: RegisterProviderArgs) -> u64 {
-    do_register_provider(ic_cdk::caller(), provider)
-}
-
-#[update(name = "unregisterProvider")]
-#[candid_method(rename = "unregisterProvider")]
-fn unregister_provider(provider_id: u64) -> bool {
-    let caller = ic_cdk::caller();
-    do_unregister_provider(caller, is_controller(&caller), provider_id)
-}
-
-#[update(name = "updateProvider")]
-#[candid_method(rename = "updateProvider")]
-fn update_provider(provider: UpdateProviderArgs) {
-    let caller = ic_cdk::caller();
-    do_update_provider(caller, is_controller(&caller), provider)
-}
-
-#[update(name = "manageProvider", guard = "require_manage_or_controller")]
-#[candid_method(rename = "manageProvider")]
-fn manage_provider(args: ManageProviderArgs) {
-    log!(
-        INFO,
-        "[{}] Managing provider: {}",
-        ic_cdk::caller(),
-        args.provider_id
-    );
-    do_manage_provider(args)
+fn get_providers() -> Vec<Provider> {
+    PROVIDERS.with(|providers| providers.clone())
 }
 
-#[query(name = "getServiceProviderMap", guard = "require_manage_or_controller")]
+#[query(name = "getServiceProviderMap")]
 #[candid_method(query, rename = "getServiceProviderMap")]
-fn get_service_provider_map() -> Vec<(RpcService, u64)> {
-    SERVICE_PROVIDER_MAP.with(|map| {
-        map.borrow()
-            .iter()
-            .filter_map(|(k, v)| Some((k.try_into().ok()?, v)))
-            .collect()
-    })
+fn get_service_provider_map() -> Vec<(RpcService, ProviderId)> {
+    SERVICE_PROVIDER_MAP.with(|map| map.iter().map(|(k, v)| (k.clone(), *v)).collect())
 }
 
 #[query(name = "getNodesInSubnet")]
 #[candid_method(query, rename = "getNodesInSubnet")]
-async fn get_nodes_in_subnet_() -> u32 {
-    get_nodes_in_subnet()
-}
-
-#[query(name = "getAccumulatedCycleCount")]
-#[candid_method(query, rename = "getAccumulatedCycleCount")]
-fn get_accumulated_cycle_count(provider_id: u64) -> u128 {
-    let caller = ic_cdk::caller();
-    do_get_accumulated_cycle_count(caller, is_controller(&caller), provider_id)
+async fn get_nodes_in_subnet() -> u32 {
+    NODES_IN_SUBNET
 }
 
-#[update(name = "withdrawAccumulatedCycles")]
-#[candid_method(rename = "withdrawAccumulatedCycles")]
-async fn withdraw_accumulated_cycles(provider_id: u64, canister_id: Principal) {
-    let caller = ic_cdk::caller();
-    do_withdraw_accumulated_cycles(caller, is_controller(&caller), provider_id, canister_id).await
+#[update(name = "insertApiKeys")]
+#[candid_method(rename = "insertApiKeys")]
+async fn insert_api_keys(api_keys: Vec<(ProviderId, String)>) {
+    for (provider_id, api_key) in api_keys {
+        insert_api_key(provider_id, ApiKey(api_key));
```

Comment:
ApiKey should implement `FromStr`, since in particular not all strings are valid API keys. Anything that contains reserved or unsafe characters should return an error (and the caller here would panic). This could for example be done by using this [crate](https://crates.io/crates/urlencoding/2.1.3) and checking that
```rust
api_key == urlencoding::encode(api_key)
```
Also ok to do that in a separate PR but I would add a TODO.

Reply:
Sure; I like this pattern over the current approach of having separate validation logic. Updated the PR, although I ended up using `TryFrom<String>` to avoid unnecessary cloning.